src/crimson/osd/ops_executer.cc

```diff
@@ -723,6 +724,115 @@ version_t OpsExecuter::get_last_user_version() const
+      snapc.snaps[0] > obc->ssc->snapset.seq) {  // existing onj is old
+
+    //  clone object, the snap field is set to the seq of the SnapContext
+    //  at it's creation.
+    hobject_t coid = soid;
+    coid.snap = snapc.seq;
+
+    unsigned l;
+    for (l=1;
+         l < snapc.snaps.size() && snapc.snaps[l] > obc->ssc->snapset.seq;
+	 l++) ;
+
+    std::vector<snapid_t> snaps(l);
+    for (unsigned i=0; i<l; i++) {
+      snaps[i] = snapc.snaps[i];
+    }
+
+    // version
+    osd_op_params->at_version = pg->next_version();
+
+    //prepare clone
+    object_info_t static_snap_oi(coid);
+    object_info_t *snap_oi;
+    if (pg->is_primary()) {
+      // lookup_or_create
+      auto [c_obc, existed] =
+        pg->get_shard_services().obc_registry.get_cached_obc(
+          std::move(coid));
+      assert(!existed);
+      c_obc->obs.oi = static_snap_oi;
+      c_obc->obs.exists = true;
+      c_obc->ssc = obc->ssc;
+      c_obc->head = obc->head;
+      logger().debug("clone_obc:{}", c_obc->obs.oi);
+      clone_obc=std::move(c_obc);
+      snap_oi = &clone_obc->obs.oi;
+    } else {
+      snap_oi = &static_snap_oi;
+    }
+
+    snap_oi->version = osd_op_params->at_version;
+    snap_oi->prior_version = obc->obs.oi.version;
+    snap_oi->copy_user_bits(obc->obs.oi);
+
+    _make_clone();
+
+    delta_stats.num_objects++;
+    if (snap_oi->is_omap()) {
+      delta_stats.num_objects_omap++;
+    }
+    delta_stats.num_object_clones++;
+    // newsnapset is obc's ssc
+    obc->ssc->snapset.clones.push_back(coid.snap);
+    obc->ssc->snapset.clone_size[coid.snap] = obc->obs.oi.size;
+    obc->ssc->snapset.clone_snaps[coid.snap] = snaps;
+
+    // clone_overlap should contain an entry for each clone
+    // (an empty interval_set if there is no overlap)
+    obc->ssc->snapset.clone_overlap[coid.snap];
+    if (obc->obs.oi.size)
+      obc->ssc->snapset.clone_overlap[coid.snap].insert(0, obc->obs.oi.size);
+
+    // log clone
+    logger().debug("cloning v {} to {} v {} snaps= {} snapset={}",
+                   obc->obs.oi.version, coid,
+                   osd_op_params->at_version, snaps, obc->ssc->snapset);
+
+    log_entries.emplace_back(pg_log_entry_t::CLONE,
+                             coid, osd_op_params->at_version,
+                             obc->obs.oi.version, obc->obs.oi.user_version,
+                             osd_reqid_t(),
+			     obc->obs.oi.mtime, 0);
+    encode(snaps, log_entries.back().snaps);
+    osd_op_params->at_version.version++;
+
+    // TODO: update most recent clone_overlap and usage stats
+
+    if (snapc.seq > obc->ssc->snapset.seq) {
+       // update snapset with latest snap context
+       obc->ssc->snapset.seq = snapc.seq;
+       obc->ssc->snapset.snaps.clear();
+    }
+    logger().debug("{} {} done, snapset={}",
+      __func__, soid, obc->ssc->snapset);
+  }
+}
+
+void OpsExecuter::_make_clone()
+{
+  logger().debug("{}", __func__);
+  ++num_write;
```

Comment:
How about using `do_write_op()` for this? The adventage is saying: _"this is a write-like op"_.